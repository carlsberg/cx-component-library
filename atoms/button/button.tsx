import React from 'react';

export type ButtonProps = React.HTMLAttributes<HTMLButtonElement>;

export const Button = ({ children, ...rest }: ButtonProps) => <button {...rest}>{children}</button>;
