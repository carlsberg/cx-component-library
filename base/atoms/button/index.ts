export { Button } from './button';
export type { ButtonInterface } from './button.types';
