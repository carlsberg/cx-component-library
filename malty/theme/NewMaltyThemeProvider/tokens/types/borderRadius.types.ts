export type PrimitiveBorderRadiusTokens = {
  square: string;
  rounded: string;
  round: string;
};

export type SemanticBorderRadiusTokens = {
  radius: string;
};
