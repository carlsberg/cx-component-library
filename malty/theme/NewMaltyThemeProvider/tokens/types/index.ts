export type { PrimitiveBorderRadiusTokens, SemanticBorderRadiusTokens } from './borderRadius.types';
export type { PrimitiveBorderTokens, SemanticBorderTokens } from './borders.types';
export type { PrimitiveColorsTokens, SemanticColorsTokens } from './colors.types';
export type { GridTokens } from './grid.types';
export type { OpacityTokens } from './opacity.types';
export type { ShadowTokens } from './shadows.types';
export type {
  AdditionalPrimitives,
  BorderRadiusTokens,
  ColorTokens,
  SemanticOverrides,
  SemanticTokens
} from './shared.types';
export type { SizeTokens } from './sizes.types';
export type { TypographyTokens } from './typography.types';
