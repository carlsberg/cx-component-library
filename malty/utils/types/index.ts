export interface SpaceProps {
  m?: string;
  mt?: string;
  mr?: string;
  mb?: string;
  ml?: string;
  mx?: string;
  my?: string;
}
