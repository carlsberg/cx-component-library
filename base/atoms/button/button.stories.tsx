import { Story } from '@storybook/react';
import React from 'react';
import { NamesTypes } from '../icon/icon.types';
import { Button } from './button';
import { ButtonInterface, ButtonType, SizeTypes } from './button.types';

export default {
  title: 'Atoms/Button',
  component: Button,
  argTypes: {
    backgroundColor: { control: 'color' },
    inverseColor: { control: 'boolean' },
    // loading: { control: 'boolean' },
    // error: { control: 'boolean' },
    // success: { control: 'boolean' },
    buttonType: {
      options: ButtonType,
      control: {
        type: 'radio'
      }
    },
    size: {
      options: SizeTypes,
      description: 'Size options are',
      table: {
        defaultValue: {
          summary: 'Medium'
        }
      },
      control: {
        type: 'radio'
      }
    }
  }
};

const Template: Story<ButtonInterface> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  buttonType: ButtonType.Primary,
  isWhite: false,
  disabled: false,
  text: 'Primary',
  size: SizeTypes.Medium
  // loading: false,
  // error: false,
  // success: false
};

export const Secondary = Template.bind({});
Secondary.args = {
  buttonType: ButtonType.Secondary,
  isWhite: false,
  disabled: false,
  text: 'Secondary',
  icon: NamesTypes.ArrowSmallUp,
  size: SizeTypes.Medium
  // loading: false,
  // error: false,
  // success: false
};

export const Floater = Template.bind({});
Floater.args = {
  buttonType: ButtonType.Floater,
  isWhite: false,
  disabled: false,
  icon: NamesTypes.Close,
  size: SizeTypes.Medium
  // loading: false,
  // error: false,
  // success: false
};

export const Link = Template.bind({});
Link.args = {
  buttonType: ButtonType.Link,
  isWhite: false,
  disabled: false,
  text: 'Link text',
  icon: NamesTypes.ArrowSmallUp,
  size: SizeTypes.Medium
  // loading: false,
  // error: false,
  // success: false
};
