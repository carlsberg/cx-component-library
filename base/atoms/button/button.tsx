import React, { useEffect, useState } from 'react';
import styled, { css, ThemeProvider } from 'styled-components';
import { ColorsTypes, NamesTypes, Sizes as IconSizes, SizesTypes as IconSizesTypes } from '../icon/icon.types';
import { theme as componentTheme } from '../theme';
import { ButtonInterface, ButtonType, Sizes, SizeTypes } from './button.types';

const IconWrapper = styled.div<{ hasText: boolean }>`
  margin-left: ${({ hasText }) => (hasText ? '10px' : 0)};
`;

const StyledIconPlaceHolder = styled.div<{ hasText: boolean }>`
  display: inline-block;
  height: ${() => IconSizes[IconSizesTypes.Medium]}px;
  width: ${() => IconSizes[IconSizesTypes.Medium]}px;
  margin-left: ${({ hasText }) => (hasText ? '10px' : 0)};
`;

const StyledButton = styled.button<{
  buttonType: ButtonType;
  hasText: boolean;
  hasIcon: boolean;
  sizing: number;
  backgroundColor: string;
  textColor: string;
  borderColor: string;
}>`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: ${({ buttonType }) => (buttonType !== ButtonType.Link ? '0 40px' : '0')};
  height: ${({ sizing, buttonType }) => (buttonType !== ButtonType.Link ? `${sizing}px` : 'auto')};
  font-size: 14px;
  font-weight: bold;
  transition: 0.25s ease-in-out;
  transition-property: background-color, color;
  cursor: pointer;
  background-color: ${({ buttonType, backgroundColor }) =>
    buttonType !== ButtonType.Link ? backgroundColor : 'transparent'};
  color: ${({ textColor }) => textColor};
  font-family: ${({ theme }) => theme.fontFamily.text};
  border: ${({ borderColor, buttonType }) => (buttonType !== ButtonType.Link ? `1px solid ${borderColor}` : 'none')};
  text-decoration: ${({ buttonType }) => (buttonType === ButtonType.Link ? 'underline' : 'none')};
  border-radius: ${({ buttonType, sizing }) => (buttonType === ButtonType.Floater ? `${sizing / 2}px` : '0')};

  &:hover {
    background-color: ${({ theme, buttonType }) =>
      buttonType !== ButtonType.Link ? theme.colors.support : theme.colors.transparent};
  }

  &:hover,
  &:focus {
    outline: 0;
  }

  &:disabled {
    cursor: default;
    color: ${({ theme }) => theme.colors.white};
    background-color: ${({ theme }) => theme.colors.disabled};
  }

  ${({ hasText, hasIcon, sizing }) =>
    !hasText &&
    hasIcon &&
    css`
      padding: 0;
      justify-content: center;
      width: ${sizing}px;
    `};
`;

function loadIconComponent(name: NamesTypes, iconColor: ColorsTypes, hasText: boolean) {
  const IconComponent = React.lazy(() => import(`../icon/icons/${name}`));
  return (
    <React.Suspense fallback={<StyledIconPlaceHolder hasText={hasText} />}>
      <IconWrapper hasText={hasText}>
        <IconComponent color={iconColor} size={SizeTypes.Small} />
      </IconWrapper>
    </React.Suspense>
  );
}

export const Button = ({
  text,
  buttonType,
  inverseColor,
  disabled,
  onClick,
  icon,
  size,
  loading,
  error,
  success
}: ButtonInterface) => {
  const [currentIcon, setCurrentIcon] = useState(icon);
  useEffect(() => {
    let iconName = icon;
    if (loading) {
      iconName = NamesTypes.Loading;
    } else if (error) {
      iconName = NamesTypes.Alert;
    } else if (success) {
      iconName = NamesTypes.CheckboxCheckSquare;
    }
    setCurrentIcon(iconName);
  }, [icon, loading, error, success]);

  const backgroundColor = inverseColor ? componentTheme.colors.white : componentTheme.colors.primary;
  const textColor = inverseColor ? componentTheme.colors.primary : componentTheme.colors.white;
  const iconColor = inverseColor ? ColorsTypes.Primary : ColorsTypes.White;
  const borderColor = inverseColor ? componentTheme.colors.primary : componentTheme.colors.transparent;

  return (
    <ThemeProvider theme={componentTheme}>
      <StyledButton
        disabled={disabled}
        hasText={!!text}
        hasIcon={!!icon}
        buttonType={buttonType}
        sizing={Sizes[size || SizeTypes.Medium]}
        backgroundColor={backgroundColor}
        textColor={textColor}
        borderColor={borderColor}
        onClick={onClick}
      >
        {text} {currentIcon && loadIconComponent(currentIcon, iconColor, !!text)}
      </StyledButton>
    </ThemeProvider>
  );
};
