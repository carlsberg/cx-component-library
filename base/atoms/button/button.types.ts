import { NamesTypes } from '../icon/icon.types';

export interface ButtonInterface extends React.HTMLAttributes<HTMLButtonElement> {
  text?: string;
  icon?: NamesTypes;
  url?: string;
  buttonType: ButtonType;
  size?: SizeTypes;
  inverseColor?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  selected?: boolean;
  disabled?: boolean;
  loading?: boolean;
  error?: boolean;
  success?: boolean;
}

export enum SizeTypes {
  Small = 'Small',
  Medium = 'Medium',
  Large = 'Large'
}

export enum Sizes {
  Small = 38,
  Medium = 48,
  Large = 58
}

export enum ButtonType {
  Primary = 'primary',
  Secondary = 'secondary',
  Floater = 'floater',
  Link = 'link'
}
