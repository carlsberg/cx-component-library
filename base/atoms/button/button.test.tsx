import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import { Button } from './button';
import { ButtonType } from './button.types';

const defaultText = 'Submit';
const newText = 'Go';

describe('button', () => {
  it('renders with correct text', () => {
    const { rerender } = render(<Button text={defaultText} buttonType={ButtonType.Primary} />);
    expect(screen.getByText(defaultText)).not.toBeNull();

    rerender(<Button text={newText} buttonType={ButtonType.Primary} />);
    expect(screen.getByText(newText)).not.toBeNull();
  });

  it('calls function on click', () => {
    const onClick = jest.fn();
    render(<Button text={defaultText} buttonType={ButtonType.Primary} onClick={onClick} />);
    fireEvent.click(screen.getByText(defaultText));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('does not call function on click when disabled', () => {
    const onClick = jest.fn();
    render(<Button text={defaultText} buttonType={ButtonType.Primary} onClick={onClick} disabled />);
    fireEvent.click(screen.getByText(defaultText));
    expect(onClick).toHaveBeenCalledTimes(0);
  });

  it('renders correctly', () => {
    const onClick = jest.fn();
    const tree = renderer
      .create(<Button text={defaultText} buttonType={ButtonType.Primary} onClick={onClick} disabled />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
