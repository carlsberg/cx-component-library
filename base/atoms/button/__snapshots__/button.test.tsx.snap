// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button renders correctly 1`] = `
<button
  className="sc-hKFxyN sc-eCApnc feAKsB jQLyOZ"
  disabled={true}
  onClick={[MockFunction]}
>
  Submit
</button>
`;
